### PR TITLE
nonsensitive no longer produces error when applied to values that are not sensitive

### LINKED ISSUE #394: nonsensitive should not error-our if value already sensitive
### OpenTF Version

```shell
1.5.x
```

### Use Cases

Pls see [PR](https://github.com/opentffoundation/opentf/pull/369) 

### Attempted Solutions

Pls see [PR](https://github.com/opentffoundation/opentf/pull/369) 

### Proposal

Make nonsensitive silently accept a value that is already non-sensitive

### References

Pls see [PR](https://github.com/opentffoundation/opentf/pull/369)

### PR DESCRIPTION
**Pull request by joaocc**
Fixes #394

This is has also been posted to hashicorp/terraform, under PR 33856. Evidence of number of users affected is mentioned there.

The behaviour of nonsensitive in 1.5.x causes an error if nonsensitive is called on a value that is not marked as "sensitive".
The documentation at https://developer.hashicorp.com/terraform/language/functions/nonsensitive offers a rationale
> nonsensitive will return an error if you pass a value that isn't marked as sensitive, because such a call would be redundant and potentially confusing or misleading to a future maintainer of your module.

Other issues describe a set of situations where users are effectively locked into a situation where they can't use derivations of collections of objects where some fields are marked as sensitive into for_each, and are also unable to use nonsensitive because some of the values are indeed nonsensitive already. It also applies to arrays of elements where some are marked as sensitive while others are not.
In these hybrid situations, there doesn't seem to be an way to mark everything as nonsensitive.

Also, this behaviour is not consistent with that of sensitive, where sensitive can be applied to any value, regardless of whether it is already sensitive or not.

However, from the sheer amount of issues complaining about the implications of this decision, this well-intentioned decision is causing unintended pain.
The current behaviour is not only interfering with developer decision of determining what is and what isn't sensitive, but also doing that when there is not even a security risk (fields are already nonsensitive). The workarounds proposed force developers away from simpler and more straightforward solutions, more difficult to understand and maintain, without any tangible benefit in terms of security.

For this reason, this PR proposes that nonsensitive does what it is meant to do, which is to allow developers to mark fields as nonsensitive, leaving softer aspects of maintainability and maintainer confusion to be decided by each developer.

This PR should not cause breaking behaviours, as it is enabling a behaviour that wasn't allowed before.

## Target Release
1.5.x (hashicorp/terraform version equivalent)

## Draft CHANGELOG entry
- nonsensitive no longer produces error when applied to values that are not sensitive
BUG FIXES
- nonsensitive can now be used to remove sensitive from any element, regardless of it's original sensitivity flag.

### CODE REVIEW

**Review by eranelbaz — COMMENTED** [review]
Hey @joaocc  
Thank you for your contribution! 
Just a quick reminder for the next pull request
- Please first open an issue, 
- Seek community approval, you can use our [community slack](https://join.slack.com/t/opentfcommunity/shared_invite/zt-22ifsm1t2-AF6cL0cOdzivP8E~4deDJA) for that
- Then submit your pull request
We try to do so to keep better collaboration between members of the community. 
Your efforts are greatly appreciated!

For more, please read our [contribution guide](https://github.com/opentffoundation/opentf/blob/main/CONTRIBUTING.md)

Please also alter the `changelog.md` file with the changes, 
And can you also write `.tf` file that explains what didn't work and how you can use this feature?
Also note that DCO check is failing, so please fix that
Thank you very much!

**Comment by marcinwyszynski** [comment]
From a purely personal perspective, I think this is valid, my intuition is that this should be a no-op. The consistency argument is very strong as well. I support merging it.

**Comment by joaocc** [comment]
[eranelbaz](https://github.com/eranelbaz), is it ok to link the original tickets in terraform repo? the tests fail due to lack of those, but I have seen elsewhere that it was discouraged to link to them directly. Thx

**Comment by eranelbaz** [comment]
@joaocc The pr linting tests?
they are failing as you need to reference an issue from our repo
So please create one and link it :)
Thank you very much for the support

**Review by eranelbaz — COMMENTED** [review]
Can you also update the `CAHNGELOG.md` file?

**Review by cube2222 — COMMENTED** [review]
Thanks for the contribution!

Also support merging this.

**Comment by eranelbaz** [comment]
Hey @joaocc 
We are putting this PR on hold as we want to get the alpha release first, so we will wait until then.

But before we can approve, can you please alter the `CHANGELOG.md` file with the change? as it changes the functionality of that operation?

**Comment by joaocc** [comment]
@eranelbaz I just did a rebase of the branch to remove the conflicts. Please proceed when suitable for you and let me know if anything else is needed. Thx

**Comment by eranelbaz** [comment]
Sure @joaocc, thanks for understanding 🙏

File: website/docs/language/functions/nonsensitive.mdx
Comment (cube2222) [inline website/docs/language/functions/nonsensitive.mdx:76]: I think we should not reference the past in docs this way ("no longer") - for a person who hasn't used this before, it doesn't make much sense.

I'd propos just saying "will not return".

**Comment by cube2222** [comment]
@joaocc Hey, sorry for asking you to do that again, but could you please rebase onto the newest main (specifically, there are changelog conflicts as we've reset it after the 1.6 release candidate).

I've also posted a nit comment to the docs.

**Comment by joaocc** [comment]
Sure. Pls take a look to check if things are as you wanted. Not sure about the updates to changelog though. Thx

**Comment by janosdebugs** [comment]
@joaocc looks good commit-wise, please allow for a few days to approve this as most of the team is not available over the holidays.

File: website/docs/language/functions/nonsensitive.mdx
Comment (cam72cam) [reply website/docs/language/functions/nonsensitive.mdx:76]: Agreed, this is the only thing I see to address before this is merged.

**Review by cube2222 — APPROVED** [review]

**Comment by cube2222** [comment]
Thanks for the contribution!

**Pull request merged**
